orchagent/pfcwdorch.cpp

```diff
@@ -282,6 +290,176 @@ void PfcWdSwOrch<DropHandler, ForwardHandler>::createEntry(const string& key,
+        string countersKey = COUNTERS_TABLE ":" + sai_serialize_object_id(queueId);
+        redisClient.hdel(countersKey, "BIG_RED_SWITCH_MODE");
+    }
+
+    m_brsEntryMap.clear();
+}
+
+template <typename DropHandler, typename ForwardHandler>
+void PfcWdSwOrch<DropHandler, ForwardHandler>::enableBigRedSwitchMode()
+{
+    SWSS_LOG_ENTER();
+
+    m_bigRedSwitchFlag =  true;
+    // Write to database that each queue enables BIG_RED_SWITCH
+    auto allPorts = gPortsOrch->getAllPorts();
+    sai_attribute_t attr;
+    attr.id = SAI_PORT_ATTR_PRIORITY_FLOW_CONTROL;
+
+    for (auto &it: allPorts)
+    {
+        Port port = it.second;
+
+        if (port.m_alias != "Ethernet8")
+        {
+            continue;
+        }
+
+        if (port.m_type != Port::PHY)
+        {
+            SWSS_LOG_INFO("Skip non-phy port %s", port.m_alias.c_str());
+            continue;
+        }
+
+        // use portorch api to get lossless tc in future.
+        sai_status_t status = sai_port_api->get_port_attribute(port.m_port_id, 1, &attr);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to get PFC mask on port %s: %d", port.m_alias.c_str(), status);
+            return;
+        }
+
+        uint8_t pfcMask = attr.value.u8;
+        for (uint8_t i = 0; i < PFC_WD_TC_MAX; i++)
+        {
+            sai_object_id_t queueId = port.m_queue_ids[i];
+            if ((pfcMask & (1 << i)) == 0 && m_entryMap.find(queueId) == m_entryMap.end())
+            {
+                continue;
+            }
+
+            string queueIdStr = sai_serialize_object_id(queueId);
+
+            vector<FieldValueTuple> countersFieldValues;
+            countersFieldValues.emplace_back("BIG_RED_SWITCH_MODE", "enable");
+            PfcWdOrch<DropHandler, ForwardHandler>::getCountersTable()->set(queueIdStr, countersFieldValues);
+        }
+    }
+
+    // Disable pfcwdaction handler on each queue if exists.
+    for (auto & entry: m_entryMap)
+    {
+        if (entry.second.handler != nullptr)
+        {
+            entry.second.handler->commitCounters();
+            entry.second.handler = nullptr;
+        }
+    }
+
+    // Create pfcwdaction hanlder on all the ports.
+    for (auto & it: allPorts)
+    {
+        Port port = it.second;
+        if (port.m_alias != "Ethernet8")
+        {
+            continue;
+        }
+
+        if (port.m_type != Port::PHY)
+        {
+            SWSS_LOG_INFO("Skip non-phy port %s", port.m_alias.c_str());
+            continue;
+        }
+
+        // use portorch api to get lossless tc in future.
+        sai_status_t status = sai_port_api->get_port_attribute(port.m_port_id, 1, &attr);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to get PFC mask on port %s: %d", port.m_alias.c_str(), status);
+            return;
+        }
+
+        uint8_t pfcMask = attr.value.u8;
```

Comment:
This is not aligned with asymmetric PFC

Reply:
This need to be updated once asym PFC is supported. I added one comment to update this part once asy is enabled.

Reply:
Oh, I thought it already was. My bad.